### PR TITLE
Add support for enums

### LINKED ISSUE #19: Enums are not supported in LogEvent.AppendGeneric
Do you have any plans regarding enumerations support? Any static cache or anything else?

### PR DESCRIPTION
**Pull request by ltrzesniewski**
This adds support for enums. Closes #19.

## New public APIs

| Member | Remarks |
| --- | --- |
| `ILogEvent.AppendEnum<T>(T) where T : Enum` | Should we just name this thing `Append<T>` ? |
| `LogManager.RegisterEnum(Type)` | |
| `LogManager.RegisterEnum<T>()` | Helper method |
| `LogManager.RegisterAllEnumsFrom(Assembly)` | Helper method |
| `LogManager.LazyRegisterEnums` | I'm not sure if we should expose this this way |
| `ZeroLogConfiguration.LazyRegisterEnums` | |

## Behavior

- By default, an enum type needs to be registered in order to be formatted correctly. Registration incurs some allocations.
- Unregistered enums will be formatted as their integer value and *may* in some unusual cases incur some allocations the first time a given type is encountered, but less than what a full registration would require.
- An optional "lazy registration" mode is provided, where trying to log an unregisteted enum will register it automatically.

## Notes

- This also changes NuGet targets to netstandard2.0, net452 and net47.
- Adds Fody for extra magic.
- Includes some refactoring and cleanup.

### CODE REVIEW

File: src/ZeroLog/EnumArg.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        public EnumArg(IntPtr typeHandle, ulong value)
+        {
+            _typeHandle = typeHandle;
+            _value = value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AppendTo(StringBuffer stringBuffer)
+        {
+            var enumString = EnumCache.TryGetString(_typeHandle, _value, out var enumRegistered);
```
Comment (MendelMonteiro) [inline src/ZeroLog/EnumArg.cs]: Shouldn't it return bool and the string as out intead?

File: src/ZeroLog/EnumCache.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            [CanBeNull]
+            public abstract string TryGetString(ulong value);
+        }
+
+        private sealed class ArrayEnumStrings : EnumStrings
+        {
+            private readonly string[] _strings;
+
+#if NET452
+            private static readonly string[] _emptyStrings = new string[0];
```
Comment (MendelMonteiro) [inline src/ZeroLog/EnumCache.cs]: This just creates noise in the code and leaves as much (or more) code to be migrated when we no longer support .NET 4.5

File: src/ZeroLog/EnumArg.cs (hunk 1 above)
Comment (ltrzesniewski) [reply src/ZeroLog/EnumArg.cs]: Not really: when `enumRegistered` is true, it doesn't necessarily mean the string won't be null. It's an additional information. The true method output is the string, and if it's null, the out parameter gives additional information about the reason why.

File: src/ZeroLog/EnumCache.cs (hunk 2 above)
Comment (ltrzesniewski) [reply src/ZeroLog/EnumCache.cs]: Yes, that's unfortunate. I'll reimplement `Array.Empty<T>` as an internal utility.

File: src/ZeroLog/EnumArg.cs (hunk 1 above)
Comment (MendelMonteiro) [reply src/ZeroLog/EnumArg.cs]: In which case the name should probably be `GetString` as the convention around `TryGetXXX` is that it returns a bool.

**Pull request merged**